File: java/src/DataStructures/BinarySearchTree.java
package DataStructures;

public class BinarySearchTree {
	
	public Node root;
	
	private static class Node {
		public int data;
		private Node left;
		private Node right;
		public Node(int data, Node left, Node right) {
			this.data = data;
			this.left = left;
			this.right = right;
		}
	}
	
	public BinarySearchTree() {
		this.root = null;
	}
	
	/**
	 * insert()
	 * O(log(n))
	 * 
	 * @param data - value to insert
	 */
	public void insert(int data) {
		root = insert(root, data);
	}
	
	public Node insert(Node n, int data) {
		if (n == null){
			n = new Node(data, null, null);
		}
		else {
			if (n.data > data) {
				n.left = insert(n.left, data);
			}
			else {
				n.right = insert(n.right, data);
			}
		}
		return n;
		
	}
	
	public void delete(int data) {
		
	}
	
	/**
	 * min(Node)
	 * Find smallest child node
	 * @param n
	 * @return
	 */
	public Node min(Node n) {
		if (n.left == null) return n;
		else return min(n.left);
	}
	
	/**
	 * max(Node)
	 * Find smallest child node
	 * @param n
	 * @return
	 */
	public Node max(Node n) {
		if (n.right == null) return n;
		else return max(n.left);
	}
	
	public void inOrder(Node n){
		if (n != null){
			inOrder(n.left);
			System.out.print(n.data + " ");
			inOrder(n.right);
		}
	}
	
	public void preOrder(Node n){
		if (n != null){
			System.out.print(n.data + " ");
			preOrder(n.left);
			preOrder(n.right);
		}
	}
	
	public void postOrder(Node n){
		if (n != null){
			postOrder(n.left);
			postOrder(n.right);
			System.out.print(n.data + " ");
		}
	}
	
	public static void main (String[] args) {
		BinarySearchTree bst = new BinarySearchTree();
		bst.insert(4);
		bst.insert(9);
		bst.insert(1);
		bst.insert(2);
		bst.insert(5);
		bst.insert(7);
		bst.insert(8);
		
		System.out.println("In-Order:");
		bst.inOrder(bst.root);
		System.out.println("\nPre-Order:");
		bst.preOrder(bst.root);
		System.out.println("\nPost-Order:");
		bst.postOrder(bst.root);
	}
}

File: python/DataStructures/BinarySearchTree.py
"""
BinarySearchTree.py

"""

class Node:
	def __init__(self, value, left=None, right=None):
		self.value = value
		self.left = left
		self.right = right

class BinarySearchTree:
	
	def __init__(self):
		self.root = None

	def get(self, value):
		"""
		Return node with given value
		"""
		return self.find(self.root, value)

	def find(self, N, value):
		if N is None: return None
		if N.value == value:
			return N
		elif N.value >= value:
			return self.find(N.left, value)
		elif N.value < value:
			return self.find(N.right, value)
		else: return None

	def insert(self, value):
		"""
		Insert value at root of tree
		"""
		self.root = self.put(self.root, value)

	def put(self, N, value):
		"""
		Insert value under given node.
		- Recursive implementation.

		best O(log(n))
		worst O(n)
		"""
		if N is None:
			N = Node(value)
		else:
			if N.value >= value:
				N.left = self.put(N.left, value)
			else:
				N.right = self.put(N.right, value)

		return N

	def delete(self, N, value):
		"""
		Delete value from tree
		"""
		if N is None: return None
		
		if N.value > value:
			N.left = self.delete(N.left, value)
		elif N.value < value:
			N.right = self.delete(N.right, value)
		else:
			if N.left is None:
				return N.right
			elif N.right is None:
				return N.left
			else:
				replace = self.min(N.right)
				N.data = replace.data
				N.right = self.delete(N.right, replace.data)
		return N

	def min(self, N):
		"""
		Find min node in the right subtree
		"""
		if N.left is None:
			return N
		else: 
			return self.min(N.left)

	def min_nonrecursive(self, N):
		while N.left is not None:
			N = N.left
		return N


	def dft_inorder(self, N):
		if N is not None:
			self.dft_inorder(N.left)
			print(N.value, end=" ")
			self.dft_inorder(N.right)

	def dft_preorder(self, N):
		if N is not None:
			print(N.value, end=" ")
			self.dft_preorder(N.left)
			self.dft_preorder(N.right)

	def dft_postorder(self, N):
		if N is not None:
			self.dft_postorder(N.left)
			self.dft_postorder(N.right)
			print(N.value, end=" ")

if __name__ == '__main__':
	BST = BinarySearchTree()
	BST.insert(4)
	BST.insert(9)
	BST.insert(1)
	BST.insert(2)
	BST.insert(5)
	BST.insert(7)
	BST.insert(8)

	print("In Order:")
	BST.dft_inorder(BST.root)
	print("\nPre Order:")
	BST.dft_preorder(BST.root)
	print("\nPost Order:")
	BST.dft_postorder(BST.root)
	print()

	BST.root = BST.delete(BST.root, 1)
	print("\n----- DELETE 1 -----")
	print("In Order:")
	BST.dft_inorder(BST.root)
	print("\nPre Order:")
	BST.dft_preorder(BST.root)
	print("\nPost Order:")
	BST.dft_postorder(BST.root)
	print()
